drivers/counter/counter_mchp_xec.c

```diff
@@ -0,0 +1,413 @@
+
+	LOG_DBG("Counter alarm set to %u ticks", ticks);
+
+	counter->control_reg.start = 1;
+
+	return 0;
+}
+
+
+static int counter_xec_cancel_alarm(struct device *dev, u8_t chan_id)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+	struct counter_xec_data *data = COUNTER_XEC_DATA(dev);
+
+	counter->control_reg.start = 0;
+	counter->int_enable_reg.int_enable = 0;
+
+	data->alarm_cb = NULL;
+	data->user_data = NULL;
+
+	LOG_DBG("Counter alarm canceled");
+
+	return 0;
+}
+
+static u32_t counter_xec_get_pending_int(struct device *dev)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+
+	return counter->status_reg.event_int;
+}
+
+static u32_t counter_xec_get_top_value(struct device *dev)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+	const struct counter_xec_config *counter_cfg = COUNTER_XEC_CONFIG(dev);
+
+	if (counter_cfg->info.max_top_value == UINT16_MAX) {
+		return (u32_t)counter->preload16_reg;
+	}
+
+	return counter->preload32_reg;
+}
+
+static int counter_xec_set_top_value(struct device *dev,
+				     const struct counter_top_cfg *cfg)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+	const struct counter_xec_config *counter_cfg = COUNTER_XEC_CONFIG(dev);
+	struct counter_xec_data *data = COUNTER_XEC_DATA(dev);
+
+	if (counter->control_reg.start || data->alarm_cb) {
+		return -EBUSY;
+	}
+
+	if (cfg->ticks != counter_cfg->info.max_top_value) {
+		return -ENOTSUP;
+	}
+
+	if (counter_cfg->info.max_top_value == UINT16_MAX) {
+		counter->preload16_reg = cfg->ticks;
+	} else {
+		counter->preload32_reg = cfg->ticks;
+	}
+
+	counter->count_reg = cfg->ticks;
+
+	data->top_cb = cfg->callback;
+	data->user_data = cfg->user_data;
+
+	if (data->top_cb) {
+		counter->int_enable_reg.int_enable = 1;
+		counter->control_reg.auto_restart = 1;
+	} else {
+		counter->int_enable_reg.int_enable = 0;
+		counter->control_reg.auto_restart = 0;
+	}
+
+	LOG_DBG("Counter top value was set to %u", cfg->ticks);
+
+	return 0;
+}
+
+static u32_t counter_xec_get_max_relative_alarm(struct device *dev)
+{
+	const struct counter_xec_config *counter_cfg = COUNTER_XEC_CONFIG(dev);
+
+	return counter_cfg->info.max_top_value;
+}
+
+static void counter_xec_isr(struct device *dev)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+	const struct counter_xec_config *counter_cfg = COUNTER_XEC_CONFIG(dev);
+	struct counter_xec_data *data = COUNTER_XEC_DATA(dev);
+	u32_t girq_result;
+
+	girq_result = MCHP_GIRQ_RESULT(counter_cfg->girq_id);
+	REG32(MCHP_GIRQ_SRC_ADDR(counter_cfg->girq_id)) = girq_result;
+
+	if (!(girq_result & counter_cfg->girq_src)) {
+		return;
+	}
+
+	LOG_DBG("Counter ISR");
+
+	if (data->alarm_cb) {
+		u32_t ticks;
+
+		if (counter_cfg->info.max_top_value == UINT16_MAX) {
+			ticks = counter->preload16_reg;
+		} else {
+			ticks = counter->preload32_reg;
+		}
+
```

Comment:
it would be better to ticks from `count_reg` since we want to return tick value that was requested and not when counter handle that. This can be used to create periodic alarm without accumulative error.

Reply:
Do you mean it is possible to call this function even if the counter has started already? set_top_value() can be called one after another?

Because current device is not continuously incrementing the counter unless it has been started. So you call set_top_value() start() ... and stop() to stop it. Then you can call again set_top_value().

Reply:
yes, in theory top value can be changed without stopping the counter and then based on flag presence counter is reset or not.